Use zeitgeist to keep track of dependencies across files

jhrozek opened the pull request:
The zeitgeist utility is being used by many Kubernetes projects to
verify external dependencies from a single `dependencies.yaml` file. We
now add an additional `make verify-dependencies` step to our CI to check
the versions via regular expressions within various locations within
this repository.

rhmdnd commented on dependencies.yaml:
```diff
@@ -0,0 +1,32 @@
+    version: v4.15.2
+    refPaths:
+    - path: Makefile
+      match: MIGRATE_VERSION
```
Cool! We can also use this to track our dependency on an AWS client and terraform, right?

jhrozek replied:
So are you saying we should expand these to contain versions that we could pin with dependencies.yaml?
btw this is a good use-case for zeitgeist (keeping docs and scripts in sync) but I'm not sure what exactly are you asking me to do..

JAORMX replied:
If you're interested in doing more extensive dependency updates. Renovate supports terraform, Dockerfiles and many other things, it's what I've been rolling out in our repos as it's more extensive than dependabot.

rhmdnd replied:
> So are you saying we should expand these to contain versions that we could pin with dependencies.yaml?
> btw this is a good use-case for zeitgeist (keeping docs and scripts in sync) but I'm not sure what exactly are you asking me to do..

I was curious if we could use something to track our dependency on Terraform.

Thoughts on Renovate versus zeitgeist? I don't have extensive experience with either of them, so I'm curious to hear both of your opinions.

rhmdnd commented on Makefile:
```diff
@@ -67,6 +67,13 @@ $(TOOLS_DIR)/golangci-lint:
+	$(TOOLS_DIR)/zeitgeist validate --local-only --base-path . --config dependencies.yaml
+
+$(TOOLS_DIR)/zeitgeist: $(TOOLS_DIR)
+	go build -o tools ./vendor/sigs.k8s.io/zeitgeist
```
This makes sense to me, so long as something installs zeitgeist locally and makes sure it's copied to the vendor directory.

Is that a manual process? I checked this change out locally to rebase it, but `go mod tidy` removes zeitgeist, and `go mod vendor` removes it from the vendors directory (causing this to fail).

I presume `go mod` is removing this dependency since it's not actually used in the source.

jhrozek replied:
`gitignore` strikes again!
We had `tools` added to the gitignore list, which was causing the `pkg/tools/tools.go` file to be ignored which was in turn causing zeitgeist, which was defined in tools to be removed..

rhmdnd replied:
/facepalm

Good catch, thanks!

jhrozek committed:
gitignore: Only ignore top-level tools/ directory, not any tools directory

jhrozek committed:
Use zeitgeist to keep track of dependencies across files

The zeitgeist utility is being used by many Kubernetes projects to
verify external dependencies from a single `dependencies.yaml` file. We
now add an additional `make verify-dependencies` step to our CI to check
the versions via regular expressions within various locations within
this repository.

rhmdnd reviewed (commented):
Looks good! I'm curious about using renovate versus zeitgeist. Are there obvious advantages to one over the other?

jhrozek commented:
On Fri, Jul 01, 2022 at 11:26:18AM -0700, Lance Bragstad wrote:
> @rhmdnd commented on this pull request.
> 
> 
> 
> > +    version: 1.46.2
> +    refPaths:
> +    - path: Makefile
> +      match: GOLANGCI_LINT_VERSION
> +
> +  - name: kubectl
> +    version: 1.24.1
> +    refPaths:
> +    - path: Makefile
> +      match: KUBECTL_VERSION
> +
> +  - name: migrate
> +    version: v4.15.2
> +    refPaths:
> +    - path: Makefile
> +      match: MIGRATE_VERSION
> 
> > So are you saying we should expand these to contain versions that we could pin with dependencies.yaml?
> > btw this is a good use-case for zeitgeist (keeping docs and scripts in sync) but I'm not sure what exactly are you asking me to do..
> 
> I was curious if we could use something to track our dependency on Terraform.
> 
> Thoughts on Renovate versus zeitgeist? I don't have extensive experience with either of them, so I'm curious to hear both of your opinions.

Honestly I picked zeitgeist and dependabot because we use them in SPO
(and dependabot because it's github's inhouse dependency updated).

If @JAORMX has some links where I could read up on renovate, I'd be
happy to learn more. I'm really not married to either zeitgeist or
dependabot, so I'm fine with changing to whatever it's better.

In the short term, I wanted to keep the dependencies current so that we
don't end up in a similar situation like with CO.
